### PR TITLE
[GLUTEN-7572][CORE] Check if iterator has been closed

### LINKED ISSUE #7572: [CORE] An error occurs when calling hasNext after ColumnarBatchOutIterator is closed
### Description

I encountered a `Module not found in resource map: 4` error in gluten 1.2.0 with uniffle. After debugging, I found that it was caused by calling `hasNext` after the `ColumnarBatchOutIterator` was closed.
It happened in `VeloxUniffleColumnarShuffleWriter`:
https://github.com/apache/incubator-gluten/blob/519176b7aaa3e75bf7be7918323b7e15f5b335ee/gluten-uniffle/velox/src/main/java/org/apache/spark/shuffle/writer/VeloxUniffleColumnarShuffleWriter.java#L129
https://github.com/apache/incubator-gluten/blob/519176b7aaa3e75bf7be7918323b7e15f5b335ee/gluten-uniffle/velox/src/main/java/org/apache/spark/shuffle/writer/VeloxUniffleColumnarShuffleWriter.java#L135

Reproduce sql:

```
select  i_item_id
       ,i_item_desc
       ,i_current_price
 from item, inventory, date_dim, store_sales
 where i_current_price between 49 and 49+30
 and inv_item_sk = i_item_sk
 and d_date_sk=inv_date_sk
 and d_date between cast('2001-01-28' as date) and date_add(cast('2001-01-28' as date) ,  60 )
 and i_manufact_id in (80,675,292,17)
 and inv_quantity_on_hand between 100 and 500
 and ss_item_sk = i_item_sk
 group by i_item_id,i_item_desc,i_current_price
 order by i_item_id
   limit 100;
```

Call Stack:
`close` ColumnarBatchOutIterator:
```
	at org.apache.gluten.vectorized.ColumnarBatchOutIterator.closeInternal(ColumnarBatchOutIterator.java:91)
	at org.apache.gluten.vectorized.GeneralOutIterator.close(GeneralOutIterator.java:59)
	at org.apache.gluten.execution.VeloxAppendBatchesExec.$anonfun$doExecuteColumnar$6(VeloxAppendBatchesExec.scala:86)
	at org.apache.gluten.utils.iterator.IteratorsV1$IteratorCompleter.tryComplete(IteratorsV1.scala:86)
	at org.apache.gluten.utils.iterator.IteratorsV1$IteratorCompleter.hasNext(IteratorsV1.scala:73)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
	at org.apache.spark.shuffle.writer.VeloxUniffleColumnarShuffleWriter.writeImpl(VeloxUniffleColumnarShuffleWriter.java:129)
	at org.apache.spark.shuffle.writer.RssShuffleWriter.write(RssShuffleWriter.java:280)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:104)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:54)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
	at org.apache.spark.scheduler.Task.run(Task.scala:141)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:620)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:623)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

`hasNext` after `ColumnarBatchOutIterator` is closed:
```
	at org.apache.gluten.vectorized.GeneralOutIterator.hasNext(GeneralOutIterator.java:39)
	at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:45)
	at org.apache.gluten.utils.iterator.IteratorsV1$ReadTimeAccumulator.hasNext(IteratorsV1.scala:127)
	at org.apache.gluten.utils.iterator.IteratorsV1$PayloadCloser.hasNext(IteratorsV1.scala:37)
	at org.apache.gluten.utils.iterator.IteratorsV1$IteratorCompleter.hasNext(IteratorsV1.scala:71)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
	at org.apache.spark.shuffle.writer.VeloxUniffleColumnarShuffleWriter.writeImpl(VeloxUniffleColumnarShuffleWriter.java:135)
	at org.apache.spark.shuffle.writer.RssShuffleWriter.write(RssShuffleWriter.java:280)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:104)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:54)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
	at org.apache.spark.scheduler.Task.run(Task.scala:141)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:620)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:623)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

### PR DESCRIPTION
**Pull request by wForget**
## What changes were proposed in this pull request?

Check if closed in `ColumnarBatchOutIterator.hasNext`

closes #7572

## How was this patch tested?

minor fix

### CODE REVIEW

**Review by zhouyuan — APPROVED** [review]

**Comment by zhztheplayer** [comment]
My intuition is when an iterator was closed, it should not be called through `hasNext()` or `next()`. Perhaps there is a root cause of https://github.com/apache/incubator-gluten/issues/7572 waiting us to find out.

Instead of the PR's change, I think we can add a sanity check in `hasNext0` and `next0` to check if the iterator is closed. If it's closed, we throw an error indicating it's an illegal state.

And we can then continue looking for a more proper solution for https://github.com/apache/incubator-gluten/issues/7572.

**Comment by wForget** [comment]
> My intuition is when an iterator was closed, it should not be called through `hasNext()` or `next()`. Perhaps there is a root cause of #7572 waiting us to find out.
> 
> Instead of the PR's change, I think we can add a sanity check in `hasNext0` and `next0` to check if the iterator is closed. If it's closed, we throw an error indicating it's an illegal state.
> 
> And we can then continue looking for a more proper solution for #7572.

Actually the bug mentioned in #7572 has been fixed by #6698, and this PR aims to avoid similar problems.

File: gluten-arrow/src/main/java/org/apache/gluten/vectorized/ColumnarBatchOutIterator.java [hunk 1]
```diff
@@ -54,6 +54,9 @@ public long itrHandle() {
 
   @Override
   public boolean hasNext0() throws IOException {
+    if (closed.get()) {
+      return false;
+    }
```
Comment (zhztheplayer) [inline gluten-arrow/src/main/java/org/apache/gluten/vectorized/ColumnarBatchOutIterator.java:59]: Not meant to be nitpicking, but I think this is not needed for now?

`hasNext() == false` literally means the iterator was drained internally while `closed == true` could mean the iterator was closed by a caller. The two are semantically different so I think we can add this kind of logic when it's really needed.

Comment (wForget) [reply gluten-arrow/src/main/java/org/apache/gluten/vectorized/ColumnarBatchOutIterator.java:59]: Makes sense, I will remove this, thank you.

Comment (wForget) [reply gluten-arrow/src/main/java/org/apache/gluten/vectorized/ColumnarBatchOutIterator.java:59]: But after closed, the native iterator has been removed. Does this mean that this iterator was drained?

**Review by zhztheplayer — APPROVED** [review]

**Comment by wForget** [comment]
> So can we close #7572 now?

yes

**Pull request merged**